webapp/channels/src/components/analytics/doughnut_chart.tsx

```diff
@@ -14,76 +13,57 @@ type Props = {
     data?: ChartData;
 }
 
-export default class DoughnutChart extends React.PureComponent<Props> {
-    private canvasRef = React.createRef<HTMLCanvasElement>();
+const DoughnutChart: React.FC<Props> = ({title, width, height, data}) => {
+    const canvasRef = useRef<HTMLCanvasElement | null>(null);
+    const chartRef = useRef<Chart<'doughnut'> | null>(null);
 
-    public chart: Chart<'doughnut'> | null = null;
-
-    public componentDidMount(): void {
-        this.initChart();
-    }
-
-    public componentDidUpdate(prevProps: Props): void {
-        if (!deepEqual(prevProps.data, this.props.data)) {
-            this.initChart(true);
-        }
-    }
-
-    public componentWillUnmount(): void {
-        if (this.chart && this.canvasRef.current) {
-            this.chart.destroy();
-        }
-    }
-
-    public initChart = (update?: boolean): void => {
-        if (typeof this.props.data === 'undefined') {
-            return;
-        }
-
-        if (!this.canvasRef.current) {
+    useEffect(() => {
```

Comment:
There are a couple of things that this effect is not doing exactly as it used to do.

In the original code, if no chart was created during mount, no chart was created at all. I find that a bug, so let's keep that change in (the change you did with substituting the `update` argument by `chartRef.current`).

Also, you are destroying the chart on every render, which causes creating a new chart on every render. We should move that part of the effect to a new effect. Something like this.
```
useEffect(() => {
    return () => {
        chartRef.current?.destroy();
    }
}, []);
```

Finally, now you are checking the `data` object, but the idea was to check the deepEqual. To achieve that, you need to store the previous data, and keep a secondary state to see whether they are equal. It is annoying, but if we need the deep equality, we have to do it right.

Something like this should work, but I am happy if you explore more elegant ideas.
```
const prevData = useRef(data);
const isEqual = data === prevData.current || deepEqual(prevData.current, data);
useEffect(() => {
    if (isEqual) {
        return;
    }
    //init/update logic
}, [data]);

useEffect(() => {
    return () => {
        prevProps.current = data;
    }
}, [data]);
```

Reply:
> In the original code, if no chart was created during mount, no chart was created at all. I find that a bug, so let's keep that change in (the change you did with substituting the update argument by chartRef.current).
Thanks for looking at it. I thought it will be bypassed.

> Something like this should work, but I am happy if you explore more elegant ideas.
I'm trying, but the block of code provided by you works great. After few tries, if I didn't get anything. I'll go with yours.

Reply:
I made a new commit `example 1`, lemme know if that logic works. Else I'll go with yours!